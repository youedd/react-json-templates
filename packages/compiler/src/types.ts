import type { ParserPlugin } from '@babel/parser'

export type RJTType = 'Serializable'
export type RJTComponentType = {
  type: RJTType,
  name: string
}


export type RJTAnalyserResult =
  | {
    type: "Exports",
    exports: Record<string, RJTComponentType>
  }
  | {
    type: "Template",
    exports: null
  }

export type RJTCompilerCache = Record<string, RJTAnalyserResult>

export interface RJTCompilerConfig {
  sourceType: 'module'
  plugins: ParserPlugin[]
}
