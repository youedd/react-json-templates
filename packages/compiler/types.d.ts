declare module 'babel-plugin-remove-unused-vars' {
  const plugin: (params: any) => any
  export default plugin
}
