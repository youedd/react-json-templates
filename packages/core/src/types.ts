
export type FIXME = any

export interface RJT_COMPONENT {
  type: '__RJT_COMPONENT__'
  name: string
  props: Record<string, unknown>
}

export interface RJT_FRAGMENT {
  type: '__RJT_FRAGMENT__'
  children: RJT_ELEMENT[]
}

export interface RJT_ACTION {
  type: '__RJT_ACTION__'
  name: string
  params: unknown[]
}

export type RJT_CONSTANT_TYPE = 'number' | 'string' | 'function' | 'object' | 'array'

export interface RJT_CONSTANT<T extends RJT_CONSTANT_TYPE = RJT_CONSTANT_TYPE> {
  type: '__RJT_CONSTANT__'
  name: string
  kind: T
}

export type RJT_ELEMENT =
  | string
  | RJT_COMPONENT
  | RJT_FRAGMENT
  | RJT_ACTION
  | RJT_ELEMENT[]

export type RJT_OPERAND_TYPE = number | RJT_CONSTANT<'number'> | RJT_GENERIC_OPERATION

export type RJT_GENERIC_OPERATION =
  { type: '__RJT_OPERATION__' }
  &
  {
    operation: '+'
    operands: [RJT_OPERAND_TYPE, RJT_OPERAND_TYPE]
  }
  | {
    operation: '*'
    operands: [RJT_OPERAND_TYPE, RJT_OPERAND_TYPE]
  }
  | {
    operation: '-'
    operands: [RJT_OPERAND_TYPE, RJT_OPERAND_TYPE]
  }
  | {
    operation: '/'
    operands: [RJT_OPERAND_TYPE, RJT_OPERAND_TYPE]
  }
  | {
    operation: 'sqrt'
    operands: [RJT_OPERAND_TYPE]
  }
  | {
    operation: 'pow'
    operands: [RJT_OPERAND_TYPE, RJT_OPERAND_TYPE]
  }

export type RJT_OPERATION_TYPE = RJT_GENERIC_OPERATION['operation']

type FindByOperation<Union, Type> = Union extends { operation: Type } ? Union : never
export type RJT_OPERATION<Type extends RJT_OPERATION_TYPE> = FindByOperation<RJT_GENERIC_OPERATION, Type>

export type RJT_PROP =
  | string
  | number
  | boolean
  | { [x: string]: RJT_PROP }
  | RJT_PROP[]

export type RJT_PROPS = Record<string, RJT_PROP>
