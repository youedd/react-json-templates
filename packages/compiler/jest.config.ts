import type { Config } from 'jest'
import baseConfig from '../../jest.config.base'

const config: Config = {
  ...baseConfig,
  displayName: 'compiler'
}

export default config
