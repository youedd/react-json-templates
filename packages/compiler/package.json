{
  "name": "@react-json-template/compiler",
  "version": "0.0.1",
  "main": "index.js",
  "license": "MIT",
  "scripts": {
    "test": "jest"
  },
  "dependencies": {
    "@babel/code-frame": "^7.21.4",
    "@babel/generator": "^7.21.4",
    "@babel/parser": "^7.21.4",
    "@babel/traverse": "^7.21.4",
    "babel-plugin-remove-unused-vars": "^2.2.0"
  },
  "devDependencies": {
    "@types/babel__code-frame": "^7.0.3",
    "@types/debug": "^4.1.7",
    "debug": "^4.3.4"
  }
}