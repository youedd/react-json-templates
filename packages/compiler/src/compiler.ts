import { InternalError, InvalidSyntaxError, ParseError } from './errors'
import type { RJTCompilerConfig, RJTCompilerCache, RJTComponentType } from './types'
import { isTemplatePath, parseFile, parseString, readFile } from './utils'
import traverse, { type NodePath } from '@babel/traverse'
import generate from '@babel/generator'
import * as types from '@babel/types'
import UnusedVarsPlugin from 'babel-plugin-remove-unused-vars'
import { getIdentifierPossibleTypes } from './typeUtils'
import Path from 'path'
import { analyze } from './analyser'
import resolve from "enhanced-resolve"

const resolveModule = resolve.create.sync({
  extensions: [".tsx", ".jsx", "ts", "js"]
});

interface Config {
  filePath: string
  compilerConfig: RJTCompilerConfig
  cache: RJTCompilerCache
}

export const compile = (config: Config): string => {
  const { filePath, compilerConfig } = config
  const { ast, code } = parseFile(filePath, compilerConfig)

  validate(filePath, ast, code)
  replaceJSX(ast, code, config)
  transform(ast)

  const newCode = generate(ast).code

  return `// @ts-nocheck \n\n ${newCode}`
}

const validate = (filePath: string, ast: types.File, code: string): void => {
  const lastExpression = ast.program.body.at(-1)

  if (
    !types.isExpressionStatement(lastExpression) ||
    !(
      types.isJSXFragment(lastExpression.expression) ||
      types.isJSXElement(lastExpression.expression)
    )
  ) {
    throw new InvalidSyntaxError(filePath, {
      code,
      start: lastExpression?.loc?.start,
      end: lastExpression?.loc?.end
    })
  }
}

const replaceJSX = (ast: types.File, code: string, config: Config): void => {
  const { filePath } = config

  traverse(
    ast,
    {
      ExportAllDeclaration(path) {
        throw new InvalidSyntaxError(filePath, {
          code,
          start: path.node.loc?.start,
          end: path.node.loc?.end
        })
      },
      ExportNamedDeclaration(path) {
        throw new InvalidSyntaxError(filePath, {
          code,
          start: path.node.loc?.start,
          end: path.node.loc?.end
        })
      },
      ExportDefaultDeclaration(path) {
        throw new InvalidSyntaxError(filePath, {
          code,
          start: path.node.loc?.start,
          end: path.node.loc?.end
        })
      },
      JSXAttribute(path) {
        const value = path.get('value')
        if (value.isJSXElement() || value.isJSXFragment()) {
          path.node.value = types.jsxExpressionContainer(value.node)
        }
      },
      JSXFragment: {
        exit(path) {
          path.replaceWith(
            types.objectExpression([
              types.objectProperty(
                types.identifier('type'),
                types.stringLiteral('__RJT_FRAGMENT__')
              ),
              types.objectProperty(
                types.identifier('children'),
                types.arrayExpression(types.react.buildChildren(path.node) as types.Expression[])
              )
            ])
          )
        }
      },
      JSXElement: {
        exit(path) {
          const openingElement = path.get('openingElement')
          const name = openingElement.get('name')

          if (!name.isJSXIdentifier()) {
            throw new InvalidSyntaxError(
              filePath,
              {
                code,
                start: openingElement.node.name.loc?.start,
                end: openingElement.node.name.loc?.end,
                message: `${openingElement.node.type} tags are not supported.`
              }
            )
          }

          const tagName = name.node.name
          const componentType = getJSXIdentifierType(name, code, config)

          if (componentType === null) {
            throw new ParseError(
              filePath,
              {
                code,
                start: openingElement.node.name.loc?.start,
                end: openingElement.node.name.loc?.end,
                message: 'JSX element is neither a Template nor a Serializable.'
              }
            )
          }

          const props = buildProps(
            openingElement.node.attributes,
            path.node
          )

          if (componentType === 'Template') {
            path.replaceWith(
              types.callExpression(
                types.identifier(tagName),
                [props]
              )
            )
            return
          }

          if (componentType.type === 'Serializable') {
            path.replaceWith(
              types.objectExpression([
                types.objectProperty(
                  types.identifier('type'),
                  types.stringLiteral('__RJT_COMPONENT__')
                ),
                types.objectProperty(
                  types.identifier('name'),
                  types.stringLiteral(componentType.name)
                ),
                types.objectProperty(
                  types.identifier('props'),
                  props
                )
              ])
            )
          }
        }
      }
    }
  )
}

const transform = (ast: types.File): void => {
  let lastImportIndex = -1
  let lastTypeDeclarationIndex = -1

  const body = ast.program.body

  body.forEach(item => {
    delete item.start
    delete item.end
    delete item.loc
  })

  const move = (from: number, to: number): void => {
    const node = body.splice(from, 1)[0]
    body.splice(to, 0, node)
  }

  for (let i = 0; i < body.length; i++) {
    const node = body[i]

    if (types.isImportDeclaration(node)) {
      move(i, ++lastImportIndex)
      lastTypeDeclarationIndex++
    } else if (types.isTSTypeAliasDeclaration(node) || types.isTSInterfaceDeclaration(node)) {
      move(i, ++lastTypeDeclarationIndex)
    }
  }

  const lastStatement = body.at(-1)

  if (!types.isExpressionStatement(lastStatement)) {
    throw new InternalError("Template's last statement should be an Expression statement")
  }

  const lastStatementExpression = lastStatement.expression

  if (!types.isExpression(lastStatement.expression)) {
    throw new InternalError('Last statement should be an Expression')
  }

  const defaultExport = types.exportDefaultDeclaration(
    types.functionDeclaration(
      null,
      [types.identifier('props')],
      types.blockStatement([
        ...body.slice(lastTypeDeclarationIndex + 1, -1),
        types.returnStatement(lastStatementExpression)
      ]),
      false,
      true
    )
  )

  body.splice(lastTypeDeclarationIndex + 1, body.length - lastTypeDeclarationIndex, defaultExport)

  const visitor = UnusedVarsPlugin({ types, traverse }).visitor

  traverse(ast, visitor)
}

const getJSXIdentifierType = (
  path: NodePath<types.JSXIdentifier>,
  code: string,
  config: Config
): 'Template' | RJTComponentType | null => {
  const binding = path.scope.getBinding(path.node.name)

  if (binding == null) {
    return null
  }

  if (binding.path.isVariableDeclarator()) {
    const possibleTypes = getIdentifierPossibleTypes(path)

    if (possibleTypes[0]?.type === 'Serializable') {
      throw new ParseError(config.filePath, {
        code,
        start: path.node.loc?.start,
        end: path.node.loc?.end,
        message: 'Serializable should not be declared in a template file'
      })
    }
  }


  let filePath: string | false = false
  let key: string | null = null

  if (binding.path.parentPath?.isImportDeclaration()) {
    const parent = binding.path.parent as types.ImportDeclaration
    const dirname = Path.dirname(config.filePath)
    filePath = resolveModule(dirname, parent.source.value)
    key = "default"

    if (filePath && isTemplatePath(filePath)) {
      return 'Template'
    }
  }

  if (binding.path.isImportSpecifier()) {
    const parent = binding.path.parent as types.ImportDeclaration
    const dirname = Path.dirname(config.filePath)
    filePath = resolveModule(dirname, parent.source.value)
    const imported = binding.path.get('imported')
    key = imported.isStringLiteral()
      ? imported.node.value
      : imported.isIdentifier()
        ? imported.node.name
        : ''
  }

  if (filePath && key) {

    const code = readFile(filePath)
    const ast = parseString(code, config.compilerConfig)

    const analyzerResult = analyze({
      cache: config.cache,
      ast,
      code,
      filePath
    })

    return analyzerResult.exports?.[key] ?? null
  }

  return null
}

const buildProps = (
  attributes?: Array<types.JSXAttribute | types.JSXSpreadAttribute>,
  children?: types.JSXElement
): types.ObjectExpression => {
  const properties: Array<types.ObjectProperty | types.SpreadElement> = []

  if (attributes != null) {
    const props = attributes.map((node) => {
      if (types.isJSXSpreadAttribute(node)) {
        return types.spreadElement(node.argument)
      }

      const propertyName = types.identifier(node.name.name as string)

      if (node.value == null) {
        return types.objectProperty(
          propertyName,
          types.booleanLiteral(true)
        )
      }

      if (types.isJSXExpressionContainer(node.value)) {
        return types.objectProperty(
          propertyName,
          types.isExpression(node.value.expression)
            ? node.value.expression
            : types.nullLiteral()
        )
      }

      return types.objectProperty(
        propertyName,
        node.value
      )
    })

    properties.push(...props)
  }

  if (children != null) {
    properties.push(
      types.objectProperty(
        types.identifier('children'),
        types.arrayExpression(types.react.buildChildren(children) as types.Expression[])
      )
    )
  }

  return types.objectExpression(properties)
}
