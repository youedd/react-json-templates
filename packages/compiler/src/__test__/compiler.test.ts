import fs from 'fs'
import { compile } from '../compiler'
import { type RJTCompilerCache, type RJTCompilerConfig } from '../types'

const S = `
import {Serializable} from "react-json-template"

export const S2 = Serializable("S2", () => null)

export default Serializable("S1", () => null)
`

const compilerConfig: RJTCompilerConfig = {
  sourceType: 'module',
  plugins: ['jsx', 'typescript']
}

const _compile = (code: string, cache: RJTCompilerCache = {}): string => {
  jest.spyOn(fs, 'readFileSync').mockImplementation((path) => {
    if (path === 'filePath') {
      return code
    }

    if (path === 'filePath/S') {
      return S
    }

    return ''
  })
  return compile({ filePath: 'filePath', compilerConfig, cache })
}

describe('compiler', () => {
  afterEach(() => {
    jest.resetAllMocks()
  })

  it('should throw Invalid syntax error if last statement is not JSX expression', () => {
    const code = `
    import {Repeat} from "./Repeat"

    const x = 5;

    <Repeat count={x}>
      Test
    </Repeat>

    const y = ""
    `

    expect(() => _compile(code)).toThrow('invalid syntax')
  })

  it('should throw Invalid syntax error if export ExportNamedDeclaration found', () => {
    const code = `
    import {Repeat} from "./Repeat"

    export const x = 5;

    <Repeat count={x}>
      Test
    </Repeat>
    `

    expect(() => _compile(code)).toThrow('invalid syntax')
  })

  it('should throw Invalid syntax error if export ExportAllDeclaration found', () => {
    const code = `
    import {Repeat} from "./Repeat"
    export * from "./exports"

    const x = 5;

    <Repeat count={x}>
      Test
    </Repeat>
    `

    expect(() => _compile(code)).toThrow('invalid syntax')
  })

  it('should throw Invalid syntax error if export ExportDefaultDeclaration found', () => {
    const code = `
    import {Repeat} from "./Repeat"
    export * from "./exports"
    
    const x = 5;

    export default x;

    <Repeat count={x}>
      Test
    </Repeat>
    `

    expect(() => _compile(code)).toThrow('invalid syntax')
  })

  it('should compile simple template', () => {
    const code = `
    import S1, {S2} from "./S"
  
    type X = number
  
    const x : X = 5;
  
    <S1>
      <S2 x={x} />
    </S1>
    `

    expect(_compile(code)).toMatchSnapshot()
  })

  it('should compile template with unordered imports and types', () => {
    const code = `
    
    type X = number
    const x : X = 5;
    
    import T1 from "./T.rjt"

    interface Y { foo: string}
    const y = { foo: "bar" };
      
    <T1 x={x} y={y} />
    `

    expect(_compile(code)).toMatchSnapshot()
  })

  it('should compile template with fragment', () => {
    const code = `
    import S1, {S2} from "./S"
    
    type X = number
    const x : X = 5;
  
    interface Y { foo: string}
    const y = { foo: "bar" };
      
    <>
      <S1 x={x} />
      <S2 y={y} />
    </>
    `

    expect(_compile(code)).toMatchSnapshot()
  })

  it('should compile template with nested templates', () => {
    const code = `
    import T1 from "./T.rjt"
    import S1 from "./S"
    
    type X = number
    const x : X = 5;
  
    interface Y { foo: string}
    const y = { foo: "bar" };
      
    <T1 x={x} y={y} s={<S1 x={x} />} />
    `

    expect(_compile(code)).toMatchSnapshot()
  })

  it('should compile template with string literals ', () => {
    const code = `
    import T1 from "./T.rjt"
    import S1 from "./S"

    <T1>
      Test
      <S1 />
    </T1>
    `

    expect(_compile(code)).toMatchSnapshot()
  })

  it('should compile template with js code ', () => {
    const code = `
    import T1 from "./T.rjt"
    import S1 from "./S"

    <T1>
      <S1 x={0} />
      {[1, 2, 3].map(item => <S1 x={item} />)}
    </T1>
    `

    expect(_compile(code)).toMatchSnapshot()
  })
})
