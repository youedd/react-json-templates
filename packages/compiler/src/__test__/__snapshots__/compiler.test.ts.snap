// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compiler should compile simple template 1`] = `
"// @ts-nocheck 

 type X = number;
export default async function (props) {
  const x: X = 5;
  return {
    type: "__RJT_COMPONENT__",
    name: "S1",
    props: {
      children: [{
        type: "__RJT_COMPONENT__",
        name: "S2",
        props: {
          x: x,
          children: []
        }
      }]
    }
  };
}"
`;

exports[`compiler should compile template with fragment 1`] = `
"// @ts-nocheck 

 type X = number;
interface Y {
  foo: string;
}
export default async function (props) {
  const x: X = 5;
  const y = {
    foo: "bar"
  };
  return {
    type: "__RJT_FRAGMENT__",
    children: [{
      type: "__RJT_COMPONENT__",
      name: "S1",
      props: {
        x: x,
        children: []
      }
    }, {
      type: "__RJT_COMPONENT__",
      name: "S2",
      props: {
        y: y,
        children: []
      }
    }]
  };
}"
`;

exports[`compiler should compile template with js code  1`] = `
"// @ts-nocheck 

 import T1 from "./T.rjt";
export default async function (props) {
  return T1({
    children: [{
      type: "__RJT_COMPONENT__",
      name: "S1",
      props: {
        x: 0,
        children: []
      }
    }, [1, 2, 3].map(item => ({
      type: "__RJT_COMPONENT__",
      name: "S1",
      props: {
        x: item,
        children: []
      }
    }))]
  });
}"
`;

exports[`compiler should compile template with nested templates 1`] = `
"// @ts-nocheck 

 import T1 from "./T.rjt";
type X = number;
interface Y {
  foo: string;
}
export default async function (props) {
  const x: X = 5;
  const y = {
    foo: "bar"
  };
  return T1({
    x: x,
    y: y,
    s: {
      type: "__RJT_COMPONENT__",
      name: "S1",
      props: {
        x: x,
        children: []
      }
    },
    children: []
  });
}"
`;

exports[`compiler should compile template with string literals  1`] = `
"// @ts-nocheck 

 import T1 from "./T.rjt";
export default async function (props) {
  return T1({
    children: ["Test", {
      type: "__RJT_COMPONENT__",
      name: "S1",
      props: {
        children: []
      }
    }]
  });
}"
`;

exports[`compiler should compile template with unordered imports and types 1`] = `
"// @ts-nocheck 

 import T1 from "./T.rjt";
type X = number;
interface Y {
  foo: string;
}
export default async function () {
  const x: X = 5;
  const y = {
    foo: "bar"
  };
  return T1({
    x: x,
    y: y,
    children: []
  });
}"
`;
